File: app/src/main/java/com/quranapp/android/adapters/ADPQuranPages.java
package com.quranapp.android.adapters;

import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import static com.quranapp.android.reader_managers.ReaderParams.RecyclerItemViewType.CHAPTER_INFO;
import static com.quranapp.android.reader_managers.ReaderParams.RecyclerItemViewType.READER_FOOTER;
import static com.quranapp.android.reader_managers.ReaderParams.RecyclerItemViewType.READER_PAGE;
import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;

import com.peacedesign.android.utils.Dimen;
import com.quranapp.android.activities.ActivityReader;
import com.quranapp.android.components.quran.QuranMeta;
import com.quranapp.android.components.reader.QuranPageModel;
import com.quranapp.android.components.reader.QuranPageSectionModel;
import com.quranapp.android.utils.extensions.LayoutParamsKt;
import com.quranapp.android.views.reader.ChapterInfoCardView;
import com.quranapp.android.views.reader.QuranPageView;
import com.quranapp.android.views.reader.ReaderFooter;

import java.util.ArrayList;

public class ADPQuranPages extends RecyclerView.Adapter<ADPQuranPages.VHQuranPage> {
    private final QuranMeta.ChapterMeta mChapterInfoMeta;
    private final ArrayList<QuranPageModel> mModels;
    private final ActivityReader mActivity;

    public ADPQuranPages(ActivityReader activity, QuranMeta.ChapterMeta chapterInfoMeta, ArrayList<QuranPageModel> models) {
        mActivity = activity;
        mChapterInfoMeta = chapterInfoMeta;
        mModels = models;

        if (chapterInfoMeta != null) {
            models.add(0, new QuranPageModel().setViewType(CHAPTER_INFO));
        }
        models.add(models.size(), new QuranPageModel().setViewType(READER_FOOTER));

        for (int i = 0; i < models.size(); i++) {
            QuranPageModel model = models.get(i);
            if (model.getViewType() == READER_PAGE) {
                for (QuranPageSectionModel section : model.getSections()) {
                    section.parentIndexInAdapter = i;
                }
            }
        }

        setHasStableIds(true);
    }

    @Override
    public int getItemCount() {
        return mModels.size();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    private int getViewType(int position) {
        return mModels.get(position).getViewType();
    }

    public void highlightVerseOnScroll(int position, int chapterNo, int verseNo) {
        QuranPageModel pageModel = getPageModel(position);
        pageModel.setScrollHighlightPendingChapterNo(chapterNo);
        pageModel.setScrollHighlightPendingVerseNo(verseNo);
        notifyItemChanged(position);
    }

    @NonNull
    @Override
    public VHQuranPage onCreateViewHolder(@NonNull ViewGroup parent, int position) {
        int viewType = getViewType(position);
        View view;
        if (viewType == CHAPTER_INFO) {
            view = new ChapterInfoCardView(mActivity);
        } else if (viewType == READER_PAGE) {
            final QuranPageView quranPageView = new QuranPageView(mActivity);
            final ViewGroup.MarginLayoutParams params = new ViewGroup.MarginLayoutParams(MATCH_PARENT, WRAP_CONTENT);
            LayoutParamsKt.updateMargins(params, Dimen.dp2px(parent.getContext(), 3));
            quranPageView.setLayoutParams(params);
            view = quranPageView;
        } else if (viewType == READER_FOOTER) {
            final ReaderFooter footer = mActivity.mNavigator.readerFooter;
            footer.clearParent();
            view = footer;
        } else {
            view = new View(parent.getContext());
        }

        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params == null) {
            params = new ViewGroup.LayoutParams(MATCH_PARENT, WRAP_CONTENT);
        } else {
            params.width = MATCH_PARENT;
        }
        view.setLayoutParams(params);

        return new VHQuranPage(view);
    }

    @Override
    public void onBindViewHolder(@NonNull VHQuranPage holder, int position) {
        holder.bind(mModels.get(position));
    }

    public QuranPageModel getPageModel(int position) {
        return mModels.get(position);
    }

    class VHQuranPage extends RecyclerView.ViewHolder {

        public VHQuranPage(@NonNull View itemView) {
            super(itemView);
        }

        public void bind(QuranPageModel pageModel) {
            final int position = getBindingAdapterPosition();

            int viewType = getViewType(position);

            if (viewType == CHAPTER_INFO) {
                ((ChapterInfoCardView) itemView).setInfo(mChapterInfoMeta);
            } else if (viewType == READER_PAGE) {
                ((QuranPageView) itemView).setPageModel(pageModel);
            }
        }
    }
}

File: app/src/main/java/com/quranapp/android/adapters/ADPQuranPages.kt
package com.quranapp.android.adapters

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.peacedesign.android.utils.Dimen
import com.quranapp.android.activities.ActivityReader
import com.quranapp.android.adapters.ADPQuranPages.VHQuranPage
import com.quranapp.android.components.quran.QuranMeta.ChapterMeta
import com.quranapp.android.components.reader.QuranPageModel
import com.quranapp.android.reader_managers.ReaderParams.RecyclerItemViewType
import com.quranapp.android.utils.extensions.updateMargins
import com.quranapp.android.views.reader.ChapterInfoCardView
import com.quranapp.android.views.reader.QuranPageView

class ADPQuranPages(private val mActivity: ActivityReader, private val mChapterInfoMeta: ChapterMeta?, private val mModels: ArrayList<QuranPageModel>)
    : RecyclerView.Adapter<ADPQuranPages.VHQuranPage>() {
    init {
        if (mChapterInfoMeta != null) {
            mModels.add(0, QuranPageModel().setViewType(RecyclerItemViewType.CHAPTER_INFO))
        }
        mModels.add(mModels.size, QuranPageModel().setViewType(RecyclerItemViewType.READER_FOOTER))

        for (i in mModels.indices) {
            val model = mModels[i]
            if (model.viewType == RecyclerItemViewType.READER_PAGE) {
                for (section in model.sections) {
                    section.parentIndexInAdapter = i
                }
            }
        }

        setHasStableIds(true)
    }

    override fun getItemCount(): Int {
        return mModels.size
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }

    private fun getViewType(position: Int): Int {
        return mModels[position].viewType
    }

    fun highlightVerseOnScroll(position: Int, chapterNo: Int, verseNo: Int) {
        val pageModel = getPageModel(position)
        pageModel.scrollHighlightPendingChapterNo = chapterNo
        pageModel.scrollHighlightPendingVerseNo = verseNo
        notifyItemChanged(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, position: Int): VHQuranPage {
        val viewType = getViewType(position)
        val view: View
        when (viewType) {
            RecyclerItemViewType.CHAPTER_INFO -> {
                view = ChapterInfoCardView(mActivity)
            }
            RecyclerItemViewType.READER_PAGE -> {
                val quranPageView = QuranPageView(mActivity)
                val params = ViewGroup.MarginLayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
                params.updateMargins(Dimen.dp2px(parent.context, 3f))
                quranPageView.layoutParams = params
                view = quranPageView
            }
            RecyclerItemViewType.READER_FOOTER -> {
                val footer = mActivity.mNavigator.readerFooter
                footer.clearParent()
                view = footer
            }
            else -> {
                view = View(parent.context)
            }
        }

        var params = view.layoutParams
        if (params == null) {
            params = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        } else {
            params.width = ViewGroup.LayoutParams.MATCH_PARENT
        }
        view.layoutParams = params

        return VHQuranPage(view)
    }

    override fun onBindViewHolder(holder: VHQuranPage, position: Int) {
        holder.bind(mModels[position])
    }

    fun getPageModel(position: Int): QuranPageModel {
        return mModels[position]
    }

    inner class VHQuranPage(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(pageModel: QuranPageModel?) {
            val position = bindingAdapterPosition
            val viewType = getViewType(position)

            if (viewType == RecyclerItemViewType.CHAPTER_INFO) {
                (itemView as ChapterInfoCardView).setInfo(mChapterInfoMeta)
            } else if (viewType == RecyclerItemViewType.READER_PAGE) {
                (itemView as QuranPageView).setPageModel(pageModel)
            }
        }
    }
}
